mlflow/utils/conda.py

```diff
@@ -115,6 +115,71 @@ def _list_conda_environments(extra_env=None):
 _PIP_CACHE_DIR = "pip_cache_pkgs"
 
 
+def _create_conda_env(
+    conda_env_path,
+    conda_env_create_path,
+    project_env_name,
+    conda_extra_env_vars,
+    capture_output,
+):
+    if conda_env_path:
+        return process._exec_cmd(
+            [
+                conda_env_create_path,
+                "env",
+                "create",
+                "-n",
+                project_env_name,
+                "--file",
+                conda_env_path,
+            ],
+            extra_env=conda_extra_env_vars,
+            capture_output=capture_output,
+        )
+    else:
+        return process._exec_cmd(
+            [
+                conda_env_create_path,
+                "create",
+                "--channel",
+                "conda-forge",
+                "--yes",
+                "--override-channels",
+                "-n",
+                project_env_name,
+                "python",
+            ],
+            extra_env=conda_extra_env_vars,
+            capture_output=capture_output,
+        )
+
+
+def _create_conda_env_retry(
+    conda_env_path, conda_env_create_path, project_env_name, conda_extra_env_vars, _capture_output
+):
+    """
+    `conda env create` command can fail with `ConnectionResetError` while downloading a package or
+    fetching repository metadata. This function retries the command up to 3 times.
+    """
+    num_attempts = 3
+    for attempt in range(num_attempts):
+        try:
+            return _create_conda_env(
+                conda_env_path,
+                conda_env_create_path,
+                project_env_name,
+                conda_extra_env_vars,
+                capture_output=True,
+            )
+        except process.ShellCommandException as e:
+            if (num_attempts - attempt - 1) > 0 and "ConnectionResetError" in str(e):
+                _logger.warning(
+                    "Conda env creation failed due to ConnectionResetError. Retrying..."
+                )
+                continue
```

Comment:
I think here we should do some cleanup for broken conda env, like code:

```
            if project_env_name in _list_conda_environments(conda_extra_env_vars):
                _logger.warning(
                    "Encountered unexpected error while creating conda environment. "
                    "Removing %s.",
                    project_env_path,
                )
                process._exec_cmd(
                    [
                        conda_path,
                        "remove",
                        "--yes",
                        "--name",
                        project_env_name,
                        "--all",
                    ],
                    extra_env=conda_extra_env_vars,
                    capture_output=False,
                )
```

Reply:
I don't think we need to. This error occurs while collecting package metadata. The environment is not created yet.

Reply:
I don't think we need to do this. We already have that. See line 295.